validator/src/validator.rs

```diff
@@ -23,104 +20,92 @@ pub fn run_validator(config: Config) {
         validator.clone(),
         CommitmentConfig::processed(),
     );
-    let program = client.program(solana_ibc::ID).unwrap();
+    let program =
+        client.program(Pubkey::from_str(&config.program_id).unwrap()).unwrap();
 
-    let trie =
-        Pubkey::find_program_address(&[solana_ibc::TRIE_SEED], &solana_ibc::ID)
-            .0;
+    let trie = Pubkey::find_program_address(
+        &[solana_ibc::TRIE_SEED],
+        &Pubkey::from_str(&config.program_id).unwrap(),
+    )
+    .0;
     let chain = Pubkey::find_program_address(
         &[solana_ibc::CHAIN_SEED],
-        &solana_ibc::ID,
+        &Pubkey::from_str(&config.program_id).unwrap(),
     )
     .0;
 
-    let (_logs_subscription, receiver) = PubsubClient::logs_subscribe(
-        &config.ws_url,
-        RpcTransactionLogsFilter::Mentions(vec![config.program_id]),
-        RpcTransactionLogsConfig {
-            commitment: Some(CommitmentConfig::processed()),
-        },
-    )
-    .unwrap();
 
-    let max_retries = 5;
 
     log::info!("Validator running");
 
-    let genesis_hash = &CryptoHash::from_base64(&config.genesis_hash)
-        .expect("Invalid Genesis Hash");
-
-    // Check if there is a pending block to sign
-    let chain_account: ChainData = program.account(chain).unwrap();
-    if chain_account.pending_block().unwrap().is_some() {
-        if !chain_account
-            .pending_block()
-            .unwrap()
-            .unwrap()
-            .signers
-            .contains(&validator.pubkey().into())
-        {
-            log::info!("Found Pending block");
-            let fingerprint =
-                &chain_account.pending_block().unwrap().unwrap().fingerprint;
-            let signature = validator.sign_message(fingerprint.as_slice());
+    let max_tries = 5;
 
-            let tx = utils::submit_call(
+    loop {
+        sleep(Duration::from_secs(5));
+        let chain_account: ChainData = program.account(chain).unwrap();
+        if chain_account.pending_block().unwrap().is_some() {
+            if chain_account
+                .pending_block()
+                .unwrap()
+                .unwrap()
+                .signers
+                .get(&validator.pubkey().into())
+                .is_none()
+            {
```

Comment:
This will work better with condition flipped:

```suggestion
            if pending_block.signers.get(&validator.pubkey().into()).is_some() {
                log::info!("You have already signed the pending block");
                continue;
            }
```

and then the rest of the code doesn’t have to be indentend and it
makes control flow easier to read.